### PR TITLE
feat(issues): Implement basic issue provider for caldav

### LINKED ISSUE #626: Use caldav as task provider
## Problem Statement

I use Nextcloud deck and tasks to organize my tasks. I would like to use this data as source for tasks in super-productivity ad this would allow me to use it without having to duplicate the tasks here.

## :grey_question: Possible Solution 

Both data (Deck and Tasks) is available via CalDAV from nextcloud. Similar to the Jira- and  Gitlab-provider this could be used as a source for tasks. Caldav is a open protocol on top of webdav. Nextcloud tiself maintains a JS library for Caldav: https://github.com/nextcloud/cdav-library

## :arrow_heading_up: Describe alternatives you've considered

The only alternative I can think of is to dump the tasks regularly in a format the super-productivity can import and use that but it would be cumbersome.

## :heavy_plus_sign: Additional context

I would like to provide an initial implementation myself but cannot promise on any timeframe. If someone else wants to have a look into this I'm happy with that.

### PR DESCRIPTION
**Pull request by janLo**
# Description

This adds another issue provider to use caldav VTODOs as issue source.

## Issues Resolved

#626 

## Check List

- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.
- [ ] Transtation files have to be adjusted

### CODE REVIEW

**Comment by janLo** [comment]
The feature works in the current state and is tested against my NC instance. To make it work with a browser version one of the Workarounds from https://github.com/nextcloud/server/issues/3131 have to be applied to NC due to the lack of CORS support in NC. I hope this gets resolved in the near future. The electron version works out of the box.

The PR is currently lacking any of the translation files as I'm not aware how they're handled. I someone can explain this I'll sort them out.

Same with the documentation: just point me to the place where I can add it.

**Comment by johannesjo** [comment]
Wow! Thank you! I am looking forward to review this! I hope to get to it this Friday.

**Comment by janLo** [comment]
I'm currently working on a issue transition to complete the caldav task if a task gets marked done.

**Comment by johannesjo** [comment]
> I'm currently working on a issue transition to complete the caldav task if a task gets marked done.

As a separate PR or should I consider this one work in progress?

**Comment by janLo** [comment]
I've put this in here now. If the user ticks the issue transition feature then the caldav-tasks will set done/undone if a transition happens in super-productivity.

What is still missing is all the translation-files. So there are the T.F.... constants in the UI. Apart from that it works for me and I'm looking forward to the Review-Feedback.

**Comment by janLo** [comment]
One little "issue" is still in the implementation: For issue-updates we rely on the other caldav clients using the server to update the last-changed property. To "fix" this we could use the etag-machanism that caldav provides but as the current task-model only allows a number-value for the last-updated value and the etag can be any string this was the least-effort solution. Also most of the clients I've seen so far seem to behave correctly here.

File: src/app/features/issue/issue.service.ts [hunk 1]
```diff
@@ -2,14 +2,15 @@ import { Injectable } from '@angular/core';
 import { IssueData, IssueDataReduced, IssueProviderKey, SearchResultItem } from './issue.model';
 import { TaskAttachment } from '../tasks/task-attachment/task-attachment.model';
 import { from, merge, Observable, of, Subject, zip } from 'rxjs';
-import { GITHUB_TYPE, GITLAB_TYPE, JIRA_TYPE } from './issue.const';
+import {CALDAV_TYPE, GITHUB_TYPE, GITLAB_TYPE, JIRA_TYPE} from './issue.const';
```
Comment (johannesjo) [inline src/app/features/issue/issue.service.ts:5]:
```suggestion
import { CALDAV_TYPE, GITHUB_TYPE, GITLAB_TYPE, JIRA_TYPE } from './issue.const';
```

File: src/app/features/issue/issue.service.ts [hunk 2]
```diff
@@ -2,14 +2,15 @@ import { Injectable } from '@angular/core';
...
 import { TaskService } from '../tasks/task.service';
 import { Task } from '../tasks/task.model';
 import { IssueServiceInterface } from './issue-service-interface';
 import { JiraCommonInterfacesService } from './providers/jira/jira-common-interfaces.service';
 import { GithubCommonInterfacesService } from './providers/github/github-common-interfaces.service';
 import { switchMap } from 'rxjs/operators';
 import { GitlabCommonInterfacesService } from './providers/gitlab/gitlab-common-interfaces.service';
+import {CaldavCommonInterfacesService} from './providers/caldav/caldav-common-interfaces.service';
```
Comment (johannesjo) [inline src/app/features/issue/issue.service.ts:13]:
```suggestion
import { CaldavCommonInterfacesService } from './providers/caldav/caldav-common-interfaces.service';
```

File: src/app/features/issue/providers/caldav/caldav-client.service.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+export class CaldavClientService {
+
+  private _clientCache = new Map<string, ClientCache>();
+
+  constructor(
+    private readonly _snackService: SnackService
+  ) {
+  }
+
+  private _getXhrProvider(cfg: CaldavCfg) {
```
Comment (johannesjo) [inline src/app/features/issue/providers/caldav/caldav-client.service.ts:37]: We also have axios available as dependency, which might help to simplify things here.

File: src/app/features/issue/providers/caldav/caldav-common-interfaces.service.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    return {additionalFields: {issueLastUpdated: issueData.last_modified}, title: issueData.summary};
+  }
+
+  getById$(id: string | number, projectId: string): Observable<CaldavIssue> {
+    return this._getCfgOnce$(projectId).pipe(
+      concatMap(caldavCfg => this._caldavClientService.getById$(id, caldavCfg))
+    );
+  }
+
+  issueLink$(issueId: string | number, projectId: string): Observable<string> {
```
Comment (johannesjo) [inline src/app/features/issue/providers/caldav/caldav-common-interfaces.service.ts:37]: I have never worked with cal dav before. But is there no UI within nextcloud for accessing those? If not we might want to make issueLink$ nullable and only display the option to go to an issue if the link is available.

File: src/app/features/issue/providers/caldav/caldav-issue/caldav-issue.model.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+export type CaldavIssueReduced = Readonly<{
+  id: string;
+  completed: boolean;
+  item_url: string;
```
Comment (johannesjo) [inline src/app/features/issue/providers/caldav/caldav-issue/caldav-issue.model.ts:4]: This is not the same thing as issueUrl$ i suppose?

File: src/app/features/issue/providers/caldav/dialog-caldav-initial-setup/dialog-caldav-initial-setup.component.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  constructor(
+    @Inject(MAT_DIALOG_DATA) public data: any,
+    private _matDialogRef: MatDialogRef<DialogCaldavInitialSetupComponent>,
+  ) {
+    this.caldavCfg = this.data.caldavCfg || DEFAULT_CALDAV_CFG;
+  }
+
+  ngOnInit() {
+  }
```
Comment (johannesjo) [inline src/app/features/issue/providers/caldav/dialog-caldav-initial-setup/dialog-caldav-initial-setup.component.ts:29]:
```suggestion
```

File: src/app/features/issue/providers/caldav/dialog-caldav-initial-setup/dialog-caldav-initial-setup.component.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import {CALDAV_CONFIG_FORM, DEFAULT_CALDAV_CFG} from '../caldav.const';
+import {CaldavCfg} from '../caldav.model';
+
+@Component({
+             selector: 'dialog-caldav-initial-setup',
+             templateUrl: './dialog-caldav-initial-setup.component.html',
+             styleUrls: ['./dialog-caldav-initial-setup.component.scss'],
+             changeDetection: ChangeDetectionStrategy.OnPush
+           })
+export class DialogCaldavInitialSetupComponent implements OnInit {
```
Comment (johannesjo) [inline src/app/features/issue/providers/caldav/dialog-caldav-initial-setup/dialog-caldav-initial-setup.component.ts:15]:
```suggestion
export class DialogCaldavInitialSetupComponent {
```

**Review by johannesjo — CHANGES_REQUESTED** [review]

**Comment by johannesjo** [comment]
Very well done! Thank you very much again! The only thing I am a bit unhappy about are the new fatty dependencies :) but I understand that there might be no (good) way around them.

File: package.json [hunk 8]
```diff
@@ -135,6 +135,7 @@
     "angular-material-css-vars": "^1.1.2",
     "angular2-promise-buttons": "^4.0.10",
     "axios": "^0.21.1",
+    "cdav-library": "nextcloud/cdav-library",
```
[the remark this answers is not in the document]
Comment (janLo) [reply package.json:138]: It seems to be - I got this from teh Nextcloud-Tasks-App. The library itself does not seem to have any releases on npm.

File: package.json [hunk 9]
```diff
@@ -154,6 +155,7 @@
     "hammerjs": "^2.0.8",
     "helpful-decorators": "^2.1.0",
     "husky": "^4.2.5",
+    "ical.js": "^1.4.0",
```
[the remark this answers is not in the document]
Comment (janLo) [reply package.json:158]: It seems to be the most stable option.

File: src/app/features/issue/providers/caldav/caldav-common-interfaces.service.ts (hunk 4 above)
Comment (janLo) [reply src/app/features/issue/providers/caldav/caldav-common-interfaces.service.ts:37]: There is a UI within NC - but NC is not the only caldav server. I used davical for a long time for example.

File: src/app/features/issue/providers/caldav/caldav-client.service.ts (hunk 3 above)
Comment (janLo) [reply src/app/features/issue/providers/caldav/caldav-client.service.ts:37]: I got this xhr-provider-solution from the cdav-library repo (they've no documentation - but it was as a usage-example within an issue).  How would axios help here?

File: src/app/features/issue/providers/caldav/caldav-issue/caldav-issue.model.ts (hunk 5 above)
Comment (janLo) [reply src/app/features/issue/providers/caldav/caldav-issue/caldav-issue.model.ts:4]: Its the link to directly download the vcalendar entry. I considered to use it for updates first. But I think its not used anywhere at the moment.

**Comment by janLo** [comment]
Thank you for you feedback. I've applied your requested changes :)

I am already using it, yes. It works good so far. I checked again and couldnt really find a replacement for ical.js that looked trustworthy enough. If anyone has a suggestion I'm happy to migrate the code. Manual parsing is not really an option in my opinion. I can look into doing sth. with axios but I don't really understand how this would simplify things here. The xhr-provider is just a wrapper to injet the auth-headers for the cdav Client-Instance.

:warning:  Can you point me to how I can generate the translation files?

**Comment by johannesjo** [comment]
For the translations you just need to edit the en.json. Then you can run `yarn int` to autofill the `t.const.ts`. I generate the rest of the via google translate using babel translate, but you don't need to worry about. I can do it later on.

I also don't have an alternative for the ical stuff. It would be nice if we could strip it down to just the stuff we need, but yeah that's also hard to maintain, so the current way looks like the best compromise to me.

**Comment by janLo** [comment]
Oh no, I've edited the t.const.ts directly. Is it enough then to add the keys to the en.json?

**Comment by johannesjo** [comment]
> Oh no, I've edited the t.const.ts directly. Is it enough then to add the keys to the en.json?

No worries. You might have done a little bit of unnecessary work (sorry, I should have given you a heads up about this...), but apart from that, you just need add the translations to the file and run `yarn int` if everything is fine, there should be no chances to the `t.const.ts` apart from maybe the order.

**Comment by janLo** [comment]
I've made some last changes:
* rebase onto the current master
* provide translation strings in en.json
* make the error messages a bit more useful
* add a check if caldav is actually configured before using it

**Comment by johannesjo** [comment]
Thank you very much! I'll have a look this Friday!

**Comment by johannesjo** [comment]
I just encountered one more minor thing when going to the project settings page of an existing project:
localhost_4200_ (3)


This happens because existing projects miss the default properties for the web dav config. This can be solved by providing a little migration script inside `src/app/features/project/migrate-projects-state.util.ts`and counting up the `PROJECT_MODEL_VERSION` version by a decimal.

**Comment by janLo** [comment]
Oh, I did not notice that, sorry. I'm fine with you doing it if it's not an issue for you. I wont be able to do anything before the end of next week.

**Comment by johannesjo** [comment]
I made the change and merged everything into master. Thank you very much again @janLo !

**Comment by janLo** [comment]
Thank you!

**Pull request merged**